### PR TITLE
Plonk

### PR DESCRIPTION
**Pull request by ajgara**
# Add Plonk

## Description

Add plonk prover and verifier using strong Fiat-Shamir. Prover unit tests are cross-validated with gnark.

## Type of change

Please delete options that are not relevant.

- [x] New feature

## Checklist
- [x] Unit tests added

### CODE REVIEW

File: proving-system/plonk/src/test_utils.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        s1: Polynomial::interpolate(&domain, &s1_lagrange),
+        s2: Polynomial::interpolate(&domain, &s2_lagrange),
+        s3: Polynomial::interpolate(&domain, &s3_lagrange),
+
+        s1_lagrange: s1_lagrange,
+        s2_lagrange: s2_lagrange,
+        s3_lagrange: s3_lagrange,
+    }
+}
```
Comment (unbalancedparentheses) [inline proving-system/plonk/src/test_utils.rs:164]: add eol

File: proving_system/plonk/src/setup.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    pub s3: Polynomial<FieldElement<F>>,
+
+    pub s1_lagrange: Vec<FieldElement<F>>,
+    pub s2_lagrange: Vec<FieldElement<F>>,
+    pub s3_lagrange: Vec<FieldElement<F>>,
+}
+
+#[allow(unused)]
+pub struct VerificationKey<G1Point> {
+    pub qm_1: G1Point,
```
Comment (ilitteri) [inline proving_system/plonk/src/setup.rs:39]: What does it mean for these to have a `_1` postfix?

**Review by ilitteri — APPROVED** [review]
LGTM! Looking forward to seeing the front end!

**Review by jrchatruc — APPROVED** [review]

File: proving_system/plonk/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pub mod prover;
+pub mod setup;
+pub mod test_utils;
+pub mod verifier;
```
Comment (MauroToscano) [inline proving_system/plonk/src/lib.rs:4]: Test utils are only used in tests, so we don't need to compile it in release mode

```suggestion
pub mod verifier;
#[cfg(test)]
pub mod test_utils;
```

File: proving_system/plonk/src/setup.rs (hunk 2 above)
Comment (ilitteri) [reply proving_system/plonk/src/setup.rs:39]: Interesting, thanks!

**Pull request merged**